### PR TITLE
Add info about using sudo in check commands in Secure Sensu

### LINKED ISSUE #108: Best practices for using sudo with Sensu checks?
_From @calebhailey on May 9, 2016 19:27_

_Copied from original issue: sensu/sensu-docs#386_

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
Adds "Dynamic runtime asset paths and sudo" section in Secure Sensu.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/108

## Review Instructions
Information sources:
- https://github.com/sensu/sensu-docs/issues/108#issuecomment-790227269
- Slack discussion about this issue and a customer question
- Jef's [gist](https://gist.github.com/jspaleta/7e192a0867fc0be5fff9673699b221bb)
- Jef's [Discourse post](https://discourse.sensu.io/t/jef-practice-preserving-needed-environment-variables-when-using-sudo-with-assets/1697)

## Questions for Reviewers
1. Is "Dynamic runtime asset paths and sudo" a good name for this section? I wondered if "Sudo and check commands" or "Set up privileged checks" (or something else) might be better.
2. Does Secure Sensu seem like the right place for this information? If it does belong in Secure Sensu, it seems like we need to reference this section in other docs (like the [asset reference](https://docs.sensu.io/sensu-go/latest/plugins/assets/) and [troubleshooting](https://docs.sensu.io/sensu-go/latest/operations/maintain-sensu/troubleshoot/#dynamic-runtime-assets)).

### CODE REVIEW

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 1]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+
+{{% notice warning %}}
+**WARNING**: Before you adjust your sudo configuration, consider whether using dynamic runtime assets for privileged checks meets your security requirements.<br><br>
+Using dynamic runtime assets with sudo has important security implications.
+Rather than making a habit of using sudo for all check commands, carefully evaluate when its appropriate to escalate privileges in this way.
+{{% /notice %}}
+
+<a name="sudo-all-assets"></a>
+
+To give Sensu access to use `sudo -E` for **all** assets installed under `var/cache/sensu/sensu-agent/`, add these lines to a file in your `/etc/sudoers.d/` directory:
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:219]:
```suggestion
If necessary give give Sensu access to use `sudo -E` for **specific** assets installed under `var/cache/sensu/sensu-agent/`, add these lines to a file in your `/etc/sudoers.d/` directory:
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 2]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+
+{{< code shell >}}
+sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/**/bin/*
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:222]:
```suggestion
sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/ASSET_SHA512_SUM/bin/MY_COMMAND
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 3]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+The `**` is a [POSIX glob][8] that will match any subdirectory other than `.` and `..`.
+`NOPASSWD:SETENV:` allows the Sensu user to run commands without a password and use `sudo -E` to preserve the Sensu user’s environment.
+
+{{% notice warning %}}
+**WARNING**: Use this example with the `**` POSIX glob as a pattern for testing in non-production environments.
+If you include these lines in your sudo configuration, Sensu will run any script installed as an asset under sudo.
+With the `sudo -E` prefix added to the check command, sudo will use the environment variables meant for the Sensu check environment.
+**This opens all installed Sensu assets for privileged operation**.<br><br>
+We recommend replacing the `**` glob with the sha512 sum for the specific build of the dynamic runtime asset to [target a specific dynamic runtime asset directory](#target-a-specific-dynamic-runtime-asset-directory).
+In production, you should also replace the `bin/*.*` and `bin/*` globs in the sudo rule patterns with specific commands to further improve security.
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:236]:
```suggestion
In development, you can replace the `bin/MY_COMMAND` and `bin/MY_COMMAND.ext` globs in the sudo rule patterns with specific commands to further improve security.
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 4]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+Rather than making a habit of using sudo for all check commands, carefully evaluate when its appropriate to escalate privileges in this way.
+{{% /notice %}}
+
+<a name="sudo-all-assets"></a>
+
+To give Sensu access to use `sudo -E` for **all** assets installed under `var/cache/sensu/sensu-agent/`, add these lines to a file in your `/etc/sudoers.d/` directory:
+
+{{< code shell >}}
+sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/**/bin/*
+sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/**/bin/*.*
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:223]:
```suggestion
sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/ASSET_SHA512_SUM/bin/MY_COMMAND.ext
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 5]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+{{< /code >}}
+
+Both lines are necessary because `bin/*.*` won't match Go assets that do not end with file extension `.ext`.
+The `**` is a [POSIX glob][8] that will match any subdirectory other than `.` and `..`.
+`NOPASSWD:SETENV:` allows the Sensu user to run commands without a password and use `sudo -E` to preserve the Sensu user’s environment.
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:228]: By removing globs and using explicit examples we don't need these.

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 6]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+
+{{% notice warning %}}
+**WARNING**: Use this example with the `**` POSIX glob as a pattern for testing in non-production environments.
+If you include these lines in your sudo configuration, Sensu will run any script installed as an asset under sudo.
+With the `sudo -E` prefix added to the check command, sudo will use the environment variables meant for the Sensu check environment.
+**This opens all installed Sensu assets for privileged operation**.<br><br>
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:234]:
```suggestion
If you include these lines in your sudo configuration, Sensu will run any script installed as an asset under sudo.
With the `sudo -E` prefix added to the check command, sudo will use the environment variables meant for the Sensu check environment.
It is highly recommended that you do not configure as such, there are typically much better ways of handling this.
**This opens all installed Sensu assets for privileged operation**.<br><br>
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 7]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+We recommend replacing the `**` glob with the sha512 sum for the specific build of the dynamic runtime asset to [target a specific dynamic runtime asset directory](#target-a-specific-dynamic-runtime-asset-directory).
+In production, you should also replace the `bin/*.*` and `bin/*` globs in the sudo rule patterns with specific commands to further improve security.
+{{% /notice %}}
+
+With these changes, you can add `sudo -E` as the prefix for all asset-provided commands.
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:239]:
```suggestion
With these changes, you can add `sudo -E` as the prefix for the specific asset-provided commands that need it.
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 8]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+This will ensure that agent-provided environment variables are passed to the sudo environment and provide the expected sensu-agent operation using default agent configuration settings.
+
+### Example check with privileged check command
+
+This example check uses the `sudo -E` prefix to run a privileged check command provided by a dynamic runtime asset, [System Profile Linux][11].
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:244]:
```suggestion
This unsafe example check uses the `sudo -E` prefix to run a privileged check command provided by a dynamic runtime asset, [System Profile Linux][11].
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 9]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+The dynamic runtime asset named `sensu/system-profile-linux` is mapped into the environment variable `SENSU_SYSTEM_PROFILE_LINUX_PATH`.
+Sensu will use the asset-specific path environment variable to prefix the check command instead of editing the sudo `secure_path`.
+
+This check requires sudo rules that [allow Sensu to use `sudo -E` to run commands for all assets][14] installed under `var/cache/sensu/sensu-agent/`.
+These rules allow Sensu to run **all** installed asset commands.
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:249]:
```suggestion
These rules allow Sensu to run the specified installed asset commands.
```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 10]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+In your `/etc/sudoers.d/` file, in place of `**`, list the sha512 sum for the specific build of the dynamic runtime asset you are using.
+This effectively whitelists the asset paths you want sudo to be able to use.
+You can find the sha512 sum for the build you want to whitelist in the asset definition.
+
+For example, the following rules would allow `sudo -E` access for only the [Linux amd64 build of the System Profile Linux][13] dynamic runtime asset:
+
+{{< code shell >}}
+sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/29a6e1725f37e533cd9f6a8a17b40ef757ba5ea5dc27701c19a7fb9ac9c2bdf9f684d758d4440f278bb7fc669e52bf02efb3fc6758731f7f3e10617a2fe026cf/bin/*.*
+sensu   ALL=(ALL)       NOPASSWD:SETENV: /var/cache/sensu/sensu-agent/29a6e1725f37e533cd9f6a8a17b40ef757ba5ea5dc27701c19a7fb9ac9c2bdf9f684d758d4440f278bb7fc669e52bf02efb3fc6758731f7f3e10617a2fe026cf/bin/*
+{{< /code >}}
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:449]: We should axe this whole section, do not offer globs as a pattern then there is no need to ask people to do security measures they wont do given the alternative.

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 11]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+
+### Extend the sudo secure_path to include assets
+
+If the sudo `secure_path` option is enabled, you can extend the path to include sensu-agent assets so that you do not have to explicitly use the full path to the asset binary in your sudo calls.
+For example, this update allows you to call `sudo echo "hello world"` rather than `sudo /usr/bin/echo "hello world"`.
+This is useful for checks that need to run on different hosts or that have different builds for different hosts.
+
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:456]:
```suggestion
This is useful for checks that need to run on different hosts or that have different builds for different hosts.
However the risks are simply not worth it given how easy it is to automate a reasonable solution.

```

File: content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md [hunk 12]
```diff
@@ -203,6 +203,344 @@ However, deployments can also use the same certificates and keys for etcd peer a
...
+Add this line to your `/etc/sudoers.d/` file:
+
+{{< code shell >}}
+Defaults    secure_path = /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/var/cache/sensu/sensu-agent/**/bin
```
Comment (majormoses) [inline content/sensu-go/6.2/operations/deploy-sensu/secure-sensu.md:460]:
```suggestion
Defaults    secure_path = /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/var/cache/sensu/sensu-agent/SHA512_SUM/bin
```

**Review by majormoses — CHANGES_REQUESTED** [review]
Thank you for taking the time to write up a really difficult subject.

I stated in the linked issue this is something I can't in good conscious recommend in a article about securing sensu.

We need to change the frame of mind to be secure by default and opt out of security if that is your desire. IMO No-one is coming to this article to see how they can weaken their security posture. They can simply google how to do that and look at stack overflow.

**Comment by hillaryfraley** [comment]
Thank you @majormoses. We are going to hold on this PR pending more discussion of the concerns you describe here and in the original issue. I'll switch this to draft and add a label.

**Comment by echlebek** [comment]
@majormoses thanks for the thorough review here, I agree with your comments, we should not hand out foot guns to people when they are widely available elsewhere online.

However, I do wonder how you reckon maintaining the sudoers file will work in the presence of changing asset SHAs. Are you suggesting this would be automated? Are you doing such things right now in your own systems? It makes me wonder if some sym-linking scheme on Sensu's part would be necessary to really pull this off effectively.

**Comment by hillaryfraley** [comment]
@majormoses I'll add my thanks for your thoughtful work here. We decided this needs further consideration on our end before we publish any formal guidance. For now, I'm closing this PR and the associated issue.

**Comment by majormoses** [comment]
> @majormoses thanks for the thorough review here, I agree with your comments, we should not hand out foot guns to people when they are widely available elsewhere online.
> 
> However, I do wonder how you reckon maintaining the sudoers file will work in the presence of changing asset SHAs. Are you suggesting this would be automated? Are you doing such things right now in your own systems? It makes me wonder if some sym-linking scheme on Sensu's part would be necessary to really pull this off effectively.

Hey sorry I missed the last part, tbh in my environment we don't currently use many checks that do require sudo. The ones that do are custom. As part of the build process we know what the sha512sum will be reported on at the end, we can shove that into our sudoers file. Automation is almost always the solution to security concerns.

**Pull request closed without merging**